feat(replays): Add highlighting to Replays player

billyvg opened the pull request:
Adds the ability to highlight nodes in the replayer as it plays.
Currently used to highlight LCP if available.

Use a new canvas to draw a highlight box on top of the element as well as label it. We still need to add in some logic to either re-draw the highlight on scroll or remove it altogether if the DOM element has been removed.

Fixes https://github.com/getsentry/sentry/issues/34163

billyvg committed:
feat(replays): Add highlighting to Replays player

Adds the ability to highlighting nodes in the replayer as it plays. Currently used to highlight LCP if availble.

billyvg commented on static/app/components/tooltip.tsx:
```diff
@@ -323,7 +324,7 @@ const PositionWrapper = styled('div')`
   z-index: ${p => p.theme.zIndex.tooltip};
 `;
 
-const TooltipContent = styled(motion.div)<{
```
Didn't mean to include this

billyvg committed:
undo tooltip change

ryan953 commented on static/app/components/replays/highlightReplayPlugin.tsx:
```diff
@@ -0,0 +1,51 @@
+  }
+}
+
+export {HighlightReplayPlugin};
```
could be default export since it's the only thing coming out, and the class matches the filename.

ryan953 replied:
saw it's done in https://github.com/getsentry/sentry/pull/34116/files

ryan953 commented on static/app/components/replays/replayContext.tsx:
```diff
@@ -172,10 +175,12 @@ export function Provider({children, events, value = {}}: Props) {
         }
       }
 
+      const highlightReplayPlugin: ReplayPlugin = new HighlightReplayPlugin({highlights});
```
I'm surprised the typehint is needed here `: ReplayPlugin`

ryan953 commented on static/app/views/replays/details.tsx:
```diff
@@ -82,9 +83,16 @@ function ReplayDetails() {
+    mergedReplayEvent?.entries[0].data
+      .filter(({op, data}) => op === 'largest-contentful-paint' && data?.nodeId > 0)
+      .map(({start_timestamp, data: {nodeId}}) => [Math.floor(start_timestamp), {nodeId}])
+  );
```
This might make sense to live inside the hook or somewhere, for future expansion and tests

billyvg replied:
@ryan953 Applied CR changes to a follow-up I've been working on with canvas: https://github.com/getsentry/sentry/pull/34116

ryan953 reviewed (approved):
shipit

billyvg committed:
feat(replays): Use canvas to highlight LCP in Replays (#34116)

* feat(replays): Use canvas to highlight LCP in Replays

Use canvas to draw some rects. The current implementation is to change the background of the highlighted element, but this a bit awkward depending on the layout of the element (or if it has a visible background). Instead we should use the existing canvas to draw a highlight box on top of the element as well as label it.

* extract to fn and add tests

* improve types

* undo type change

* better type

billyvg committed:
fix(ui): Restore tooltip animations (#34140)

billyvg committed:
test types

billyvg committed:
wrong import on merge conflict
